lib/publiccloud/provider.pm

```diff
@@ -230,37 +220,87 @@ sub create_instance {
     $args{image} //= $self->get_image_id();
 
+    my @vms = $self->terraform_apply(count => '1');
+    my $instance = $vms[0];
+    record_info('INFO', "VM created '" . Dumper $instance);
+
+    my $timeout = 60;
+    while ($timeout > 0) {
+        last if script_run("nc -vz -w 1 $instance->{public_ip} 22") == 0;
+        $timeout -= 1;
+        sleep 5;
+    }
+    die('Cannot reach port 22 on IP ' . $instance->{public_ip}) if ($timeout == 0);
+    return $instance;
+}
+
+=head2 terraform_apply
+
+Calls terraform tool and applies the corresponding configuration .tf file
+
+=cut
+sub terraform_apply {
+    my ($self, %args) = @_;
+    my @instances;
+
+    $args{count}                //= '1';
+    $args{instance_type}        //= get_var('PUBLIC_CLOUD_INSTANCE_TYPE');
+    $args{image}                //= $self->get_image_id();
+    $args{ssh_private_key_file} //= '.ssh/id_rsa';
+    $args{user}                 //= $self->username;
+
     record_info('INFO', "Creating instance $args{instance_type} from $args{image} ...");
-    my $ipa = $self->ipa(
-        instance_type => $args{instance_type},
-        cleanup       => 0,
-        image_id      => $args{image}
-    );
-    return $ipa->{instance};
+    assert_script_run('terraform init', $terraform_timeout);
+    my $cmd = "terraform plan ";
+    $cmd .= "-var 'image_id=" . $args{image} . "' ";
+    $cmd .= "-var 'count=" . $args{count} . "' ";
+    $cmd .= "-var 'type=" . $args{instance_type} . "' ";
+    $cmd .= "-var 'region=" . $self->region . "' ";
+    $cmd .= "-out myplan";
+    assert_script_run($cmd);
+    assert_script_run('terraform apply myplan', $terraform_timeout);
+
+    my $vms = decode_json(script_output("terraform output -json vm_name"))->{value};
+    my $ips = decode_json(script_output("terraform output -json public_ip"))->{value};
+
+    foreach my $i (0 .. $#{$vms}) {
+        my $instance = publiccloud::instance->new(
+            public_ip    => @{$ips}[$i],
+            instance_id  => @{$vms}[$i],
+            username     => $args{user},
+            ssh_key      => $args{ssh_private_key_file},
+            image_id     => $args{image},
+            region       => $self->region,
+            type         => $args{instance_type},
+            provider     => $self,
+            created_from => 'terraform'
+        );
+        push @instances, $instance;
+    }
+    # Return an ARRAY of objects 'instance'
+    return @instances;
 }
 
-=head2 destroy_instance
+=head2 terraform_destroy
+
+Destroys the current terraform deployment
 
-Destroy a instance previously created with this provider. Require a C<publiccloud::instance> object.
 =cut
-sub destroy_instance {
-    my ($self, $instance) = @_;
-    $self->ipa(cleanup => 1, running_instance_id => $instance->instance_id);
+sub terraform_destroy {
+    my ($self) = @_;
+    record_info('INFO', 'Removing terraform plan...');
+    script_run('terraform destroy -auto-approve', $terraform_timeout);
```

Comment:
What happen, if I didn't called terraform_apply() before?

Reply:
Nothing really, terraform will tell you that there is nothing to be deleted. To be in the safe place, I am calling it with script_run without "assert".